contracts/serverless-v2/tree/TreeMapUpgradeable.sol

```diff
@@ -0,0 +1,410 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+
+/// @dev Contract implements an upgradeable version of tree which stores
+/// elements in order of insertion. When a element is added, it is added
+/// to the left most empty leaf in the tree. When an element is deleted,
+/// it is replaced with the element in the right most leaf in the tree.
+/// Each element in the tree stores the weight of all elements on the left
+/// and right side of the node.
+contract TreeMapUpgradeable is Initializable {
+    /// @notice Struct that stores the value on the node and the sum of
+    /// weights on left and right side of the node.
+    /// @param value Value on the node
+    /// @param leftSum Sum of nodes on the left side of the current node
+    /// @param rightSum Sum of nodes on the right side of the current node
+    struct Node {
+        uint64 value;
+        uint64 leftSum;
+        uint64 rightSum;
+    }
+
+    struct Tree {
+        /// @notice Mapping of address of a node to it's index in nodes array
+        mapping(address => uint256) addressToIndexMap;
+        /// @notice Mapping of index in nodes array to address of the node
+        mapping(uint256 => address) indexToAddressMap;
+        /// @notice Array of nodes stored in the tree
+        Node[] nodes;
+    }
+
+    /// @custom:storage-location erc7201:marlin.oyster.storage.EnvTree
+    struct TreeMapStorage {
+        /// @notice Tree data for each environment
+        mapping(uint8 => Tree) envTree;
+    }
+
+    // keccak256(abi.encode(uint256(keccak256("marlin.oyster.storage.TreeMap")) - 1)) & ~bytes32(uint256(0xff))
+    bytes32 private constant TreeMapStorageLocation = 0x03deb113f53cfc5c9ce28544dc6140711774fc5c00752abd22c8d30d1faf2900;
+
+    error TreeInvalidInitState();
+
+    function _getTreeMapStorage() private pure returns (TreeMapStorage storage $) {
+        assembly {
+            $.slot := TreeMapStorageLocation
+        }
+    }
+
+    function __TreeMapUpgradeable_init_unchained() internal onlyInitializing {
+    }
+
+    /// @dev Initializes the tree with 0 element as the first element.
+    /// Node indexes start from 1.
+    function _init_tree(uint8 _env) internal {
+        TreeMapStorage storage $ = _getTreeMapStorage();
+        if ($.envTree[_env].nodes.length != 0) revert TreeInvalidInitState();
+        // root starts from index 1
+        $.envTree[_env].nodes.push(Node(0, 0, 0));
+    }
+
+    /// @dev Deletes the tree storage for a given env.
+    function _delete_tree(uint8 _env) internal {
+        TreeMapStorage storage $ = _getTreeMapStorage();
+
+        uint256 len = $.envTree[_env].nodes.length;
+        for (uint256 index = 0; index < len; index++) {
```

Comment:
you can start from index == 1, because for Node 0 there won't be any index to address or address to index mapping.